### PR TITLE
Avoid accessing the Workspaec off of a Solution in features that want to know if they can make a change.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We want to move away from a world where the (mutable) Workspace is directly exposed from Solution instances.  That way we can separate out purely functional query/transform APIs, versus feature code that actually needs to mutate the state the user is operating on.

This PR makes it so that features that need to be disabled if the workspace doesn't support the type of hcnage they want to make can do so without actually accessing the Workspace off of a Solution snapshot anymore.

This also exposes a helper API for this purpose internally.  https://github.com/dotnet/roslyn/issues/63457 tracks making this a public API that we can then have others use as well.

### CODE REVIEW

File: src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs [hunk 1]
```diff
@@ -54,9 +54,11 @@ private void ExecuteCommandWorker(PasteCommandArgs args, SnapshotPoint? caretPos
                 return;
 
             var subjectBuffer = args.SubjectBuffer;
-            var workspace = subjectBuffer.GetWorkspace();
-            if (workspace == null)
+            if (!subjectBuffer.TryGetWorkspace(out var workspace) ||
+                !workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
+            {
                 return;
+            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs:61]: features (like command handlers) taht are already operating on mutable buffers are totally free to get back to their workspace to query things.  They are the point of integration with teh workspace and it does make sense for them to both query it to know what is allowed, and to directly ask it to make changes.

File: src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs [hunk 2]
```diff
@@ -65,9 +67,6 @@ private void ExecuteCommandWorker(PasteCommandArgs args, SnapshotPoint? caretPos
             if (!_globalOptions.GetOption(FormattingOptionsMetadata.FormatOnPaste, document.Project.Language))
                 return;
 
-            if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
-                return;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs:69]: this just moved up.

File: src/EditorFeatures/Core/InlineRename/AbstractEditorInlineRenameService.SymbolRenameInfo.cs [hunk 3]
```diff
@@ -211,7 +211,7 @@ public bool TryOnAfterGlobalSymbolRenamed(Workspace workspace, IEnumerable<Docum
             public InlineRenameFileRenameInfo GetFileRenameInfo()
             {
                 if (RenameSymbol.Kind == SymbolKind.NamedType &&
-                    _document.Project.Solution.Workspace.CanApplyChange(ApplyChangesKind.ChangeDocumentInfo))
+                    _document.Project.Solution.CanApplyChange(ApplyChangesKind.ChangeDocumentInfo))
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/InlineRename/AbstractEditorInlineRenameService.SymbolRenameInfo.cs:214]: query/transformation APIs that operate just on snapshots can now determine what changes the snapshot supports without going back to the workspace directly.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixesAndRefactorings/CodeFixContextExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    internal static class CodeFixContextExtensions
+    {
+        /// <summary>
+        /// Remove once https://github.com/dotnet/roslyn/issues/63457 is approved
+        /// </summary>
+        public static bool CanApplyChange(this CodeFixContext context, ApplyChangesKind kind)
+        {
+            var solution = context.Document.Project.Solution;
+#if CODE_STYLE
+            return solution.Workspace.CanApplyChange(kind);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixesAndRefactorings/CodeFixContextExtensions.cs:16]: still accessing teh workspace here.  but ideally only temporarily.  once https://github.com/dotnet/roslyn/issues/63457 is approved we'll have a public API that our own code, and the CODE_STYLE layer can use.

**Comment by CyrusNajmabadi** [comment]
@akhera99 @davidwengier ptal.  thanks!

File: src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs (hunk 1 above)
Comment (akhera99) [reply src/EditorFeatures/Core/Formatting/FormatCommandHandler.Paste.cs:61]: makes sense 👍

**Review by akhera99 — APPROVED** [review]
Looks good! 
I'm surprised there are so few instances of directly getting the workspace from the solution

**Review by genlu — APPROVED** [review]

**Pull request merged**